File: E-paper_Separate_Program/5inch_e-Paper/RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd5in0.py
# *****************************************************************************
# * | File        :	  epd5in0.py
# * | Author      :   Waveshare team
# * | Function    :   5inch e-paper 
# * | Info        :
# *----------------
# * | This version:   V1.0
# * | Date        :   2025-11-04
# # | Info        :   python demo
# -----------------------------------------------------------------------------
# ******************************************************************************/
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

import logging
from . import epdconfig

import PIL
import time
from PIL import Image
import io

# Display resolution
EPD_WIDTH       = 960
EPD_HEIGHT      = 552

GRAY1  = 0xff #white
GRAY2  = 0xC0
GRAY3  = 0x80 #gray
GRAY4  = 0x00 #Blackest

logger = logging.getLogger(__name__)

class EPD:
    def __init__(self):
        self.reset_pin = epdconfig.RST_PIN
        self.dc_pin = epdconfig.DC_PIN
        self.busy_pin = epdconfig.BUSY_PIN
        self.cs_pin = epdconfig.CS_PIN
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT
        self.GRAY1  = GRAY1 #white
        self.GRAY2  = GRAY2
        self.GRAY3  = GRAY3 #gray
        self.GRAY4  = GRAY4 #Blackest
        

    # Hardware reset
    def reset(self):
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(20) 
        epdconfig.digital_write(self.reset_pin, 0)         # module reset
        epdconfig.delay_ms(2)
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(20)   

    def send_command(self, command):
        epdconfig.digital_write(self.dc_pin, 0)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([command])
        epdconfig.digital_write(self.cs_pin, 1)

    def send_data(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([data])
        epdconfig.digital_write(self.cs_pin, 1)
        
    # send a lot of data   
    def send_data2(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte2(data)
        epdconfig.digital_write(self.cs_pin, 1)
        
    def ReadBusy(self):
        logger.debug("e-Paper busy H")
        while(epdconfig.digital_read(self.busy_pin) == 1):
            epdconfig.delay_ms(1)
        logger.debug("e-Paper busy H release")

    def TurnOnDisplay(self):
        self.send_command(0x22)  
        self.send_data(0xF7)     
        self.send_command(0x20)  
        self.ReadBusy()

    def TurnOnDisplay_Fast(self):
        self.send_command(0x22) #Display Update Control
        self.send_data(0xD7)
        self.send_command(0x20) #Activate Display Update Sequence
        self.ReadBusy()

    def TurnOnDisplay_Part(self):
        self.send_command(0x22) #Display Update Control
        self.send_data(0xFF)
        self.send_command(0x20) #Activate Display Update Sequence
        self.ReadBusy()

    def SetWindow(self, x_start, y_start, x_end, y_end):
        self.send_command(0x44) # SET_RAM_X_ADDRESS_START_END_POSITION
        self.send_data(x_start & 0xFF)
        self.send_data((x_start>>8) & 0xFF)
        self.send_data(x_end & 0xFF)
        self.send_data((x_end>>8) & 0xFF)
        
        self.send_command(0x45) # SET_RAM_Y_ADDRESS_START_END_POSITION
        self.send_data(y_start & 0xFF)
        self.send_data((y_start >> 8) & 0xFF)
        self.send_data(y_end & 0xFF)
        self.send_data((y_end >> 8) & 0xFF)

    def SetCursor(self, x, y):
        self.send_command(0x4E) # SET_RAM_X_ADDRESS_COUNTER
        # x point must be the multiple of 8 or the last 3 bits will be ignored
        self.send_data(x & 0xFF)
        self.send_data((x>>8) & 0xFF)
        
        self.send_command(0x4F) # SET_RAM_Y_ADDRESS_COUNTER
        self.send_data(y & 0xFF)
        self.send_data((y >> 8) & 0xFF)
        
    def init(self):
        if (epdconfig.module_init() != 0):
            return -1
        # EPD hardware init start
        self.reset()
        self.ReadBusy()
        self.send_command(0x12)
        self.ReadBusy()

        self.send_command(0x18)   
        self.send_data(0x80)

        self.send_command(0x0C)  
        self.send_data(0xAE)
        self.send_data(0xC7)
        self.send_data(0xC3)
        self.send_data(0xC0)
        self.send_data(0x80)

        self.send_command(0x01)  
        self.send_data((EPD_HEIGHT-1)%256)
        self.send_data((EPD_HEIGHT-1)//256)
        self.send_data(0x02)

        self.send_command(0x11)
        self.send_data(0x01)

        self.send_command(0x44)
        self.send_data(0x00)
        self.send_data(0x00)
        self.send_data((EPD_WIDTH-1)%256)
        self.send_data((EPD_WIDTH-1)//256)

        self.send_command(0x45)
        self.send_data((EPD_HEIGHT-1)%256)
        self.send_data((EPD_HEIGHT-1)//256)
        self.send_data(0x00)
        self.send_data(0x00)

        self.send_command(0x4E)
        self.send_data(0x00)
        self.send_data(0x00)
        self.send_command(0x4F)
        self.send_data(0x00)
        self.send_data(0x00)
        self.ReadBusy()

        self.send_command(0x3C)
        self.send_data(0x01)
        return 0

    def init_Fast(self):
        # EPD hardware init start
        self.reset()
        self.ReadBusy()
        self.send_command(0x12)
        self.ReadBusy()

        self.send_command(0x18)   
        self.send_data(0x80)

        self.send_command(0x0C)  
        self.send_data(0xAE)
        self.send_data(0xC7)
        self.send_data(0xC3)
        self.send_data(0xC0)
        self.send_data(0x80)

        self.send_command(0x01)  
        self.send_data((EPD_HEIGHT-1)%256)
        self.send_data((EPD_HEIGHT-1)//256)
        self.send_data(0x02)

        self.send_command(0x11)
        self.send_data(0x01)

        self.send_command(0x44)
        self.send_data(0x00)
        self.send_data(0x00)
        self.send_data((EPD_WIDTH-1)%256)
        self.send_data((EPD_WIDTH-1)//256)

        self.send_command(0x45)
        self.send_data((EPD_HEIGHT-1)%256)
        self.send_data((EPD_HEIGHT-1)//256)
        self.send_data(0x00)
        self.send_data(0x00)

        self.send_command(0x4E)
        self.send_data(0x00)
        self.send_data(0x00)
        self.send_command(0x4F)
        self.send_data(0x00)
        self.send_data(0x00)
        self.ReadBusy()

        self.send_command(0x3C)
        self.send_data(0x01)

        self.send_command(0x18)
        self.send_data(0x80)
        self.send_command(0x1A)
        self.send_data(0x6A)
        
    def getbuffer(self, image):
        # logger.debug("bufsiz = ",int(self.width/8) * self.height)
        buf = [0xFF] * (int(self.width / 8) * self.height)
        image_monocolor = image.convert('1')
        imwidth, imheight = image_monocolor.size
        pixels = image_monocolor.load()
        # logger.debug("imwidth = %d, imheight = %d",imwidth,imheight)
        if imwidth == self.width and imheight == self.height:
            logger.debug("Horizontal")
            for y in range(imheight):
                for x in range(imwidth):
                    # Set the bits for the column of pixels at the current position.
                    if pixels[x, y] == 0:
                        buf[int((x + y * self.width) / 8)] &= ~(0x80 >> (x % 8))
        elif imwidth == self.height and imheight == self.width:
            logger.debug("Vertical")
            for y in range(imheight):
                for x in range(imwidth):
                    newx = y
                    newy = self.height - x - 1
                    if pixels[x, y] == 0:
                        buf[int((newx + newy * self.width) / 8)] &= ~(0x80 >> (y % 8))
        return buf  

    def display(self, image):
        self.send_command(0x24)
        self.send_data2(image)

        self.TurnOnDisplay()
    
    def display_Base(self, image):
        self.send_command(0x24)
        self.send_data2(image)

        self.send_command(0x26)
        self.send_data2(image)

        self.TurnOnDisplay()

    def display_Fast(self, image):
        self.send_command(0x24)
        self.send_data2(image)

        self.TurnOnDisplay_Fast()

    def display_Partial(self, Image):
        
        self.reset()

        self.send_command(0x18)
        self.send_data(0x80)

        self.send_command(0x3C)
        self.send_data(0x80)

        self.SetWindow(0, self.height-1, self.width-1, 0)

        self.SetCursor(0, 0)

        self.send_command(0x24)

        self.send_data2(Image)

        self.TurnOnDisplay_Part()

    def Clear(self):
        self.send_command(0x24)
        self.send_data2([0xFF] * int(self.height) * int(self.width/8))
        self.send_command(0x26)
        self.send_data2([0x00] * int(self.height) * int(self.width/8))

        self.TurnOnDisplay()

    def sleep(self):
        self.send_command(0x10) # DEEP_SLEEP
        self.send_data(0X01)
        
        epdconfig.delay_ms(2000)
        epdconfig.module_exit()
### END OF FILE ###

File: E-paper_Separate_Program/5inch_e-Paper/ESP32/EPD_5in0.cpp
/*****************************************************************************
* | File      	:   EPD_5in0.c
* | Author      :   Waveshare team
* | Function    :   5inch e-paper 
* | Info        :
*----------------
* |	This version:   V1.0
* | Date        :   2025-11-04
* | Info        :
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
******************************************************************************/
#include "EPD_5in0.h"
#include "Debug.h"
#include <time.h> 

/******************************************************************************
function :	Software reset
parameter:
******************************************************************************/
static void EPD_5in0_Reset(void)
{
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
    DEV_Digital_Write(EPD_RST_PIN, 0);
    DEV_Delay_ms(2);
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
}

/******************************************************************************
function :	send command
parameter:
     Reg : Command register
******************************************************************************/
static void EPD_5in0_SendCommand(UBYTE Reg)
{
    DEV_Digital_Write(EPD_DC_PIN, 0);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Reg);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :	send data
parameter:
    Data : Write data
******************************************************************************/
static void EPD_5in0_SendData(UBYTE Data)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Data);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :	Wait until the busy_pin goes LOW
parameter:
******************************************************************************/
static void EPD_5in0_ReadBusy(void)
{
    Debug("e-Paper busy\r\n");
    while(DEV_Digital_Read(EPD_BUSY_PIN) == 1) {      //LOW: idle, HIGH: busy
        DEV_Delay_ms(1);
    }
    Debug("e-Paper busy release\r\n");
}

/******************************************************************************
function :	Turn On Display full
parameter:
******************************************************************************/
static void EPD_5in0_TurnOnDisplay(void)
{ 
    EPD_5in0_SendCommand(0x22);
    EPD_5in0_SendData(0xF7);
	EPD_5in0_SendCommand(0x20);
    EPD_5in0_ReadBusy();

}

static void EPD_5in0_TurnOnDisplay_Part(void)
{
    EPD_5in0_SendCommand(0x22);
    EPD_5in0_SendData(0xFF);
    EPD_5in0_SendCommand(0x20);
    EPD_5in0_ReadBusy();
}

/******************************************************************************
function :	Initialize the e-Paper register
parameter:
******************************************************************************/
void EPD_5in0_Init(void)
{
    EPD_5in0_Reset();

    EPD_5in0_ReadBusy();
    EPD_5in0_SendCommand(0x12);  //SWRESET
    EPD_5in0_ReadBusy();

    EPD_5in0_SendCommand(0x18);
    EPD_5in0_SendData(0x80);

    EPD_5in0_SendCommand(0x0C);
	EPD_5in0_SendData(0xAE);
	EPD_5in0_SendData(0xC7);
	EPD_5in0_SendData(0xC3);
	EPD_5in0_SendData(0xC0);
	EPD_5in0_SendData(0x80);   //0x80 0x20

    EPD_5in0_SendCommand(0x01); //Driver output control
    EPD_5in0_SendData((EPD_5in0_HEIGHT-1)%256);
    EPD_5in0_SendData((EPD_5in0_HEIGHT-1)/256);
    EPD_5in0_SendData(0x02);


    EPD_5in0_SendCommand(0x11); //data entry mode       
	EPD_5in0_SendData(0x01);

	EPD_5in0_SendCommand(0x44); //set Ram-X address start/end position   
	EPD_5in0_SendData(0x00);
	EPD_5in0_SendData(0x00);
	EPD_5in0_SendData((EPD_5in0_WIDTH-1)%256);    
	EPD_5in0_SendData((EPD_5in0_WIDTH-1)/256);

	EPD_5in0_SendCommand(0x45); //set Ram-Y address start/end position    
    EPD_5in0_SendData((EPD_5in0_HEIGHT-1)%256);    
	EPD_5in0_SendData((EPD_5in0_HEIGHT-1)/256);  
	EPD_5in0_SendData(0x00);
	EPD_5in0_SendData(0x00);

    EPD_5in0_SendCommand(0x4E);   // set RAM x address count to 0;
	EPD_5in0_SendData(0x00);
	EPD_5in0_SendData(0x00);
	EPD_5in0_SendCommand(0x4F);   // set RAM y address count to 0X199;    
	EPD_5in0_SendData(0x00);
	EPD_5in0_SendData(0x00);
    EPD_5in0_ReadBusy();

    EPD_5in0_SendCommand(0x3C); //BorderWavefrom
    EPD_5in0_SendData(0x01);


}

/******************************************************************************
function :	Clear screen
parameter:
******************************************************************************/
void EPD_5in0_Clear(void)
{
    UWORD Width, Height;
    Width = (EPD_5in0_WIDTH % 8 == 0)? (EPD_5in0_WIDTH / 8 ): (EPD_5in0_WIDTH / 8 + 1);
    Height = EPD_5in0_HEIGHT;

    EPD_5in0_SendCommand(0x24);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_5in0_SendData(0XFF);
        }
    }
    EPD_5in0_SendCommand(0x26);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_5in0_SendData(0XFF);
        }
    }
    EPD_5in0_TurnOnDisplay();
}

void EPD_5in0_Clear_Black(void)
{
    UWORD Width, Height;
    Width = (EPD_5in0_WIDTH % 8 == 0)? (EPD_5in0_WIDTH / 8 ): (EPD_5in0_WIDTH / 8 + 1);
    Height = EPD_5in0_HEIGHT;

    EPD_5in0_SendCommand(0x24);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_5in0_SendData(0X00);
        }
    }
    EPD_5in0_SendCommand(0x26);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_5in0_SendData(0X00);
        }
    }
    EPD_5in0_TurnOnDisplay();
}


/******************************************************************************
function :	Sends the image buffer in RAM to e-Paper and displays
parameter:
******************************************************************************/
void EPD_5in0_Display(const UBYTE *Image)
{
    UWORD Width, Height;
    Width = (EPD_5in0_WIDTH % 8 == 0)? (EPD_5in0_WIDTH / 8 ): (EPD_5in0_WIDTH / 8 + 1);
    Height = EPD_5in0_HEIGHT;

    EPD_5in0_SendCommand(0x24);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_5in0_SendData(Image[i + j * Width]);
        }
    }
    EPD_5in0_TurnOnDisplay();
}

void EPD_5in0_Display_Base(const UBYTE *Image)
{
    UWORD Width, Height;
    Width = (EPD_5in0_WIDTH % 8 == 0)? (EPD_5in0_WIDTH / 8 ): (EPD_5in0_WIDTH / 8 + 1);
    Height = EPD_5in0_HEIGHT;

    EPD_5in0_SendCommand(0x24);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_5in0_SendData(Image[i + j * Width]);
        }
    }
    
    EPD_5in0_SendCommand(0x26);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_5in0_SendData(Image[i + j * Width]);
        }
    }
    EPD_5in0_TurnOnDisplay();
}

void EPD_5in0_Display_Base_fImage(UBYTE *Image)
{
	UWORD i, j, k;
	UWORD height = EPD_5in0_HEIGHT;
	UWORD width = EPD_5in0_WIDTH/8;
	
	EPD_5in0_SendCommand(0x24);   //write RAM for black(0)/white (1)
    k = 0 ;
	for(i=0; i<height; i++)
	{
		for(j=0; j<width; j++)
        {
            if(i< height/2 & j< width/2)
                EPD_5in0_SendData(Image[k++]);
            else
                EPD_5in0_SendData(0xff);
        }
	}

	EPD_5in0_SendCommand(0x26);   //write RAM for black(0)/white (1)
	k = 0 ;
	for(i=0; i<height; i++)
	{
		for(j=0; j<width; j++)
        {
            if(i< height/2 & j< width/2)
                EPD_5in0_SendData(Image[k++]);
            else
                EPD_5in0_SendData(0xff);
        }
	}
	EPD_5in0_TurnOnDisplay();	
}

/******************************************************************************
function :	Sends the image buffer in RAM to e-Paper and displays
parameter:
******************************************************************************/
void EPD_5in0_Display_Partial(const UBYTE *Image, UWORD Xstart, UWORD Ystart, UWORD Xend, UWORD Yend)
{
	if((Xstart % 8 + Xend % 8 == 8 && Xstart % 8 > Xend % 8) || Xstart % 8 + Xend % 8 == 0 || (Xend - Xstart)%8 == 0)
    {
        Xstart = Xstart / 8 ;
        Xend = Xend / 8;
    }
    else
    {
        Xstart = Xstart / 8 ;
        Xend = Xend % 8 == 0 ? Xend / 8 : Xend / 8 + 1;
    }
    

    UWORD i, Width;
	Width = Xend -  Xstart;
	UDOUBLE IMAGE_COUNTER = Width * (Yend-Ystart);

    Xend -= 1;
	Yend -= 1;	

    EPD_5in0_Reset();

    EPD_5in0_SendCommand(0x18);
    EPD_5in0_SendData(0x80);

    EPD_5in0_SendCommand(0x3C);
    EPD_5in0_SendData(0x80);

	EPD_5in0_SendCommand(0x44);
    EPD_5in0_SendData((Xstart*8) & 0xFF);
    EPD_5in0_SendData(((Xstart*8) >> 8) & 0xFF);
    EPD_5in0_SendData((Xend*8) & 0xFF);
    EPD_5in0_SendData(((Xend*8) >> 8) & 0xFF);
	
    EPD_5in0_SendCommand(0x45);
    EPD_5in0_SendData(Yend & 0xFF);
    EPD_5in0_SendData((Yend >> 8) & 0xFF);
    EPD_5in0_SendData(Ystart & 0xFF);
    EPD_5in0_SendData((Ystart >> 8) & 0xFF);
    
    EPD_5in0_SendCommand(0x4E); 
    EPD_5in0_SendData((Xstart*8) & 0xFF);
    EPD_5in0_SendData(((Xstart*8) >> 8) & 0xFF);
	
    EPD_5in0_SendCommand(0x4F);
    EPD_5in0_SendData(Ystart & 0xFF);
    EPD_5in0_SendData((Ystart >> 8) & 0xFF);

	EPD_5in0_SendCommand(0x24);   //Write Black and White image to RAM
    for (i = 0; i < IMAGE_COUNTER; i++) {
	    EPD_5in0_SendData(Image[i]);
	}

	EPD_5in0_TurnOnDisplay_Part();
}


/******************************************************************************
function :	Enter sleep mode
parameter:
******************************************************************************/
void EPD_5in0_Sleep(void)
{
    EPD_5in0_SendCommand(0x10); //enter deep sleep
    EPD_5in0_SendData(0x01);
    DEV_Delay_ms(100);
}
